tool/tsh/mfa.go

```diff
@@ -0,0 +1,487 @@
+		return trace.Wrap(err)
+	}
+
+	var devs []*types.MFADevice
+	if err := client.RetryWithRelogin(cf.Context, tc, func() error {
+		pc, err := tc.ConnectToProxy(cf.Context)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer pc.Close()
+		aci, err := pc.ConnectToCurrentCluster(cf.Context, false)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer aci.Close()
+
+		resp, err := aci.GetMFADevices(cf.Context, &proto.GetMFADevicesRequest{})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		devs = resp.Devices
+		return nil
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	printMFADevices(devs, c.verbose)
+	return nil
+}
+
+func printMFADevices(devs []*types.MFADevice, verbose bool) {
+	if verbose {
+		t := asciitable.MakeTable([]string{"Name", "ID", "Type", "Added at", "Last used"})
+		for _, dev := range devs {
+			t.AddRow([]string{
+				dev.Metadata.Name,
+				dev.Id,
+				dev.MFAType(),
+				dev.AddedAt.Format(time.RFC1123),
+				dev.LastUsed.Format(time.RFC1123),
+			})
+		}
+		fmt.Println(t.AsBuffer().String())
+	} else {
+		t := asciitable.MakeTable([]string{"Name", "Type", "Added at", "Last used"})
+		for _, dev := range devs {
+			t.AddRow([]string{
+				dev.GetName(),
+				dev.MFAType(),
+				dev.AddedAt.Format(time.RFC1123),
+				dev.LastUsed.Format(time.RFC1123),
+			})
+		}
+		fmt.Println(t.AsBuffer().String())
+	}
+}
+
+type mfaAddCommand struct {
+	*kingpin.CmdClause
+}
+
+func (c *mfaAddCommand) run(cf *CLIConf) error {
+	devType, err := prompt.PickOne(os.Stdout, os.Stdin, "Choose device type", []string{"TOTP", "U2F"})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	var typ proto.AddMFADeviceRequestInit_DeviceType
+	switch devType {
+	case "TOTP":
+		typ = proto.AddMFADeviceRequestInit_TOTP
+	case "U2F":
+		typ = proto.AddMFADeviceRequestInit_U2F
+	default:
+		// prompt.PickOne should catch this for us.
+		return trace.BadParameter("unknown device type %q", devType)
+	}
+
+	devName, err := prompt.Input(os.Stdout, os.Stdin, "Enter device name")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	devName = strings.TrimSpace(devName)
+	if devName == "" {
+		return trace.BadParameter("device name can not be empty")
+	}
+
+	dev, err := c.addDeviceRPC(cf, devName, typ)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Printf("MFA device %q added.\n", dev.Metadata.Name)
+	return nil
+}
+
+func (c *mfaAddCommand) addDeviceRPC(cf *CLIConf, devName string, devType proto.AddMFADeviceRequestInit_DeviceType) (*types.MFADevice, error) {
+	tc, err := makeClient(cf, true)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var dev *types.MFADevice
+	if err := client.RetryWithRelogin(cf.Context, tc, func() error {
+		pc, err := tc.ConnectToProxy(cf.Context)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer pc.Close()
+		aci, err := pc.ConnectToCurrentCluster(cf.Context, false)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer aci.Close()
+
+		stream, err := aci.AddMFADevice(cf.Context)
```

Comment:
That's a good idea.
I was hesitant at first, because there are many user prompts in this flow. But maybe I'll just pass in stdin/stdout as parameters.
Left a TODO for now, I'll clean this up along with other nice-to-haves if there's free time before the deadline.